### DISCUSSION TITLE
holdApplicationUntilProxyStarts like configuration in Linkerd2 ?

### DISCUSSION THREAD

**Discussion by BlackRider97** [linkerd/linkerd2#5151] [General]
Istio has added supported for [holdApplicationUntilProxyStarts](https://istio.io/latest/news/releases/1.7.x/announcing-1.7/change-notes/), which causes the sidecar injector to inject the sidecar at the start of the pod’s container list and configures it to block the start of all other containers until the proxy is ready.

I am using [Argo ](https://github.com/argoproj/argo) workload pods along with Linkerd2 as Service Mesh technology.  I don't want to start main processing logic unless Linkerd proxy is ready to accept connection which takes few seconds only.

Another way to solve this problem is wait until proxy port is not open but this requires some Init container definition which I wanted to avoid. 

Is there any configuration to hold application until proxy is not ready ?

**Comment by kleimkuhler** [comment]
Hey @BlackRider97 it may be worth checking out [linkerd-await](https://github.com/olix0r/linkerd-await). There is not a similarly convenient config option for setting this, but it should take care of your issue.

#3450 which has been open for a while and would help make this easier to find; it may be worth coming back around to that. Let me know if this solves your issue and if you have any questions about using it.

**Reply by nesl247** [reply]
This is nice for internal applications, but isn't great for external ones, as it requires building a custom image every time. Is there any plan to solve this given that Kubernetes themselves are taking eons to solve this problem?

**Reply by kleimkuhler** [reply]
@nesl247 There has not been any further discussion about providing a solution for this does not involve `linkerd-await`. It's maybe worth exploring other solutions a little more, such as using the container startup ordering as mentioned above.

**Reply by nesl247** [reply]
@kleimkuhler As mentioned, it's not possible (without significant work) to use `linkerd-await`. We'd be happy to use the pod order, but I am not aware of any way to tell linkerd to inject the container in the first position.

**Reply by kleimkuhler** [reply]
Yep I understand the issue with `linkerd-await`. And the solution for container order would probably have to be built into Linkerd.

Unfortunately there is not that much I can suggest right now to fix this, but as I said I think a solution similar to the wait container is worth exploring more.

**Reply by wmorgan** [reply]
Could Linkerd implement something similar to Istio's `holdApplicationUntilProxyStarts` behavior?

**Reply by kleimkuhler** [reply]
@wmorgan Yep I got something working over the weekend and just posted #5967! It introduces a `config.linkerd.io/await-proxy` annotation which forces the container to wait for proxy-readiness.

**Reply by wmorgan** [reply]
@kleimkuhler Nice! We're on the same page :)
